text/0000-ide.md

```diff
@@ -0,0 +1,530 @@
+- Feature Name: n/a
+- Start Date: 2015-10-13
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+This RFC describes how we intend to modify the compiler to support IDEs. The
+intention is that support will be as generic as possible. A follow-up internals
+post will describe how we intend to focus our energies and deploy Rust support
+in actual IDEs.
+
+There are two sets of technical changes proposed in this RFC: changes to how we
+compile, and the creation of an 'oracle' tool (name of tool TBC).
+
+This RFC is fairly detailed, it is intended as a straw-man plan to guide early
+implementation, rather than as a strict blueprint.
+
+
+## Compilation model
+
+An IDE will perform two kinds of compilation - an incremental check as the user
+types (used to provide error and code completion information) and a full build.
+The full build is explicitly signaled by the user (it could also happen
+implicitly, for example when the user saves a file). A full build is basically
+just a `cargo build` command, as would be done from the command line. It will
+take advantage of any future improvements to regular compilation (such as
+incremental compilation), but there is essentially no change from a compile
+today. It is not very interesting and won't be discussed further.
+
+The incremental check follows a new model of compilation. This check must be as
+fast as possible but does not need to generate machine code. We'll describe it
+in more detail below. We call this kind of compilation a 'quick-check'.
+
+This RFC also covers making compilation more robust.
+
+
+## The oracle
+
+The oracle is a long running daemon process. It will keep a database
+representation of an entire project's source code and semantic information (as
+opposed to the compiler which operates on a crate at a time). It is
+incrementally updated by the compiler and provides an IPC API for providing
```

Comment:
*Preliminary notes: I'm commenting on this RFC because I started writing a Rust plugin for the QtCreator IDE. This plugin is in a very preliminary state and it does nothing useful for the moment. I don't know much about compilers in general, and I'm discovering the QtCreator API as I write my plugin. So, be suspicious of anything I could say!*

What is the rationale for proposing a daemon process + an IPC API? I fail to see an advantage compared to the oracle just being a library with both a Rust and a C API, that I could call directly from my IDE plugin code. It would remove the pain of writing communication code. it would also allow me to instanciate two services at the same time with isolated content.

The only pros I can see for the daemon approach are:

1. *Share information between multiple IDE instances.*
    It seems like a rare use case, and I doubt there are much data to share.

2. *Be easier to integrate in languages with awful FFI capabilities.*
    The IPC API could be designed on top of an existing library, if the need is real.

When I investigated to integrate Racer to my IDE plugin, I came to the conclusion that it would be easier to use Racer as a library rather than invoking it as a process and parsing the output.